pkg/solver/singlehost.go

```diff
@@ -0,0 +1,325 @@
+				scheme = "http"
+			} else {
+				scheme = string(endpoint.Protocol)
+			}
+
+			if scheme != "http" && scheme != "https" {
+				// we cannot expose non-http endpoints publicly, because ingresses/routes only support http(s)
+				continue
+			}
+
+			if endpoint.Secure {
+				scheme = "https"
+
+				// TODO this should also do the magic of ensuring user authentication however we are going to do it
+				// in the future
+			}
+
+			publicURLPrefix := getPublicURLPrefixForEndpoint(workspaceID, machineName, endpoint)
+
+			publicURL := scheme + "://" + ensureDoesntEndWithSlash(host) + ensureDoesntEndWithSlash(publicURLPrefix) + ensureStartsWithSlash(endpoint.Path)
+
+			attrs := map[string]string{}
+			err := endpoint.Attributes.Into(&attrs)
+			if err != nil {
+				return nil, false, err
+			}
+
+			exposedEndpoints = append(exposedEndpoints, dwo.ExposedEndpoint{
+				Name:       endpoint.Name,
+				Url:        publicURL,
+				Attributes: attrs,
+			})
+		}
+		exposed[machineName] = exposedEndpoints
+	}
+
+	return exposed, true, nil
+}
+
+func (c *CheRoutingSolver) getGatewayConfigMaps(cheManager *dwoche.CheManager, workspaceID string, routing *dwo.WorkspaceRouting) []corev1.ConfigMap {
+	restrictedAnno, setRestrictedAnno := routing.Annotations[config.WorkspaceRestrictedAccessAnnotation]
+
+	labels := defaults.GetLabelsForComponent(cheManager, "gateway-config")
+	labels[config.WorkspaceIDLabel] = workspaceID
+	if setRestrictedAnno {
+		labels[config.WorkspaceRestrictedAccessAnnotation] = restrictedAnno
+	}
+
+	configMap := corev1.ConfigMap{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      defaults.GetGatewayWorkpaceConfigMapName(workspaceID),
+			Namespace: cheManager.Namespace,
+			Labels:    labels,
+			Annotations: map[string]string{
+				defaults.ConfigAnnotationWorkspaceRoutingName:      routing.Name,
+				defaults.ConfigAnnotationWorkspaceRoutingNamespace: routing.Namespace,
+			},
+		},
+		Data: map[string]string{},
+	}
+
+	routers := ""
+	services := ""
+	middlewares := ""
+
+	for machineName, endpoints := range routing.Spec.Endpoints {
+		// we need to support unique endpoints - so 1 port can actually be accessible
+		// multiple times, each time using a different resulting external URL.
+		// non-unique endpoints are all represented using a single external URL
+		ports := map[int32]map[string]bool{}
+		for _, e := range endpoints {
+			i := int32(e.TargetPort)
+
+			name := ""
+			if e.Attributes.GetString(uniqueEndpointAttributeName, nil) == "true" {
+				name = e.Name
+			}
+
+			if ports[i] == nil {
+				ports[i] = map[string]bool{}
+			}
+
+			ports[i][name] = true
+		}
+
+		for port, names := range ports {
+			for endpointName := range names {
+				var name string
+				var prefix string
+				var serviceURL string
+
+				if endpointName == "" {
+					name = fmt.Sprintf("%s-%s-%d", workspaceID, machineName, port)
+				} else {
+					name = fmt.Sprintf("%s-%s-%d-%s", workspaceID, machineName, port, endpointName)
+				}
```

Comment:
Watch out for the 63 char limit here.

Reply:
This is the contents of the traefik config file. I don't think traefik names have the 63 chars limit but actually this is an interesting thing to verify...

Reply:
Ah sorry, I misunderstood.